File: 16. 3Sum Closest.py
class Solution:
    def threeSumClosest(self, nums: 'List[int]', target: 'int') -> 'int':
        closest = None
        closest_diff = None
        
        nums.sort()
        
        for i in range(len(nums)-2):
            l = i+1
            r = len(nums)-1
            
            while l < r:
                s = nums[i]+nums[l]+nums[r]
                
                if s==target:
                    #found the best solution
                    return target
                
                diff = s-target
                if diff < 0:
                    diff *=-1
                
                if closest == None or diff < closest_diff:                    
                    closest = s
                    closest_diff = diff
                
                if s > target-diff:
                    r-=1

                elif s < target+diff:                
                    l+=1

                else:
                    break
                       
        return closest
    
            

File: 16. 3Sum Closest.c
 #include <limits.h>

int comp(const void *a, const void *b){
    return *(int*)a - *(int*)b;
}

void dump(int *nums, int numsSize){
    printf("dump: ");
    for(int i=0; i<numsSize; i++){
        printf("%d, ", nums[i]);
    }
    printf("\n");
}

int threeSumClosest(int* nums, int numsSize, int target){

    //sort nums first
    qsort(nums, numsSize, sizeof(int), comp);
    //dump(nums, numsSize);
    
    int closest = nums[0]+nums[1]+nums[2] - target;
    int i,l,r,diff;
    for(i=0; i<numsSize-2; i++){
        l=i+1;
        r=numsSize-1;
        
        while(l<r){
            diff = nums[i]+nums[l]+nums[r] - target;
            //printf("%d,%d,%d diff=%d\n", i, l, r, diff);
            if((diff*diff)<(closest*closest)){//compare abs diff
                closest = diff;
            }
            //printf("closest=%d\n",closest);
            
            if(diff<0){
                l++;
            }else if(diff>0){
                r--;
            }else{
                //found the best solution
                break;
            }
        }
        
    }
    
    return closest+target;
}

